dask_ml/feature_extraction/text.py

```diff
@@ -0,0 +1,139 @@
+        and always treated as a token separator).
+    n_features : integer, default=(2 ** 20)
+        The number of features (columns) in the output matrices. Small numbers
+        of features are likely to cause hash collisions, but large numbers
+        will cause larger coefficient dimensions in linear learners.
+    norm : 'l1', 'l2' or None, optional
+        Norm used to normalize term vectors. None for no normalization.
+    binary : boolean, default=False.
+        If True, all non zero counts are set to 1. This is useful for discrete
+        probabilistic models that model binary events rather than integer
+        counts.
+    dtype : type, optional
+        Type of the matrix returned by fit_transform() or transform().
+    alternate_sign : boolean, optional, default True
+        When True, an alternating sign is added to the features as to
+        approximately conserve the inner product in the hashed space even for
+        small n_features. This approach is similar to sparse random projection.
+    """
+    def __init__(self, input='content', encoding='utf-8',
+                 decode_error='strict', strip_accents=None,
+                 lowercase=True, preprocessor=None, tokenizer=None,
+                 stop_words=None, token_pattern=r"(?u)\b\w\w+\b",
+                 ngram_range=(1, 1), analyzer='word', n_features=(2 ** 20),
+                 binary=False, norm='l2', alternate_sign=True,
+                 dtype=np.float64):
+        self.input = input
+        self.encoding = encoding
+        self.decode_error = decode_error
+        self.strip_accents = strip_accents
+        self.preprocessor = preprocessor
+        self.tokenizer = tokenizer
+        self.analyzer = analyzer
+        self.lowercase = lowercase
+        self.token_pattern = token_pattern
+        self.stop_words = stop_words
+        self.n_features = n_features
+        self.ngram_range = ngram_range
+        self.binary = binary
+        self.norm = norm
+        self.alternate_sign = alternate_sign
+        self.dtype = dtype
+
+    def fit(self, X, y=None):
+        """Does nothing: this transformer is stateless."""
+        # triggers a parameter validation
+
+        self._get_hasher().fit(None, y=None)
+        return self
+
+    def transform(self, X):
+        """Transform a sequence of documents to a document-term matrix.
+        Parameters
+        ----------
+        X : dask.Bag of raw text documents, length = n_samples
+            Samples. Each sample must be a text document (either bytes or
+            unicode strings, file name or file object depending on the
+            constructor argument) which will be tokenized and hashed.
+        Returns
+        -------
+        X : scipy.sparse matrix, shape = (n_samples, self.n_features)
+            Document-term matrix.
+        """
+
+        transformer = super(HashingVectorizer, self).transform
+
+        if isinstance(X, Bag):
+            return X.map_partitions(transformer)
```

Comment:
It looks like this could also be a Dask Series array or anything that implements a `map_partitions` method that applies that function over iterables.  Perhaps something like the following would work instead?

```python
if hasattr(X, 'map_partitions):
    return X.map_partitions(transformer)
```

Also, what is the desired return type here?  Do we want this to be a Dask array of sparse arrays?  My guess is that dask.bag wouldn't give that to us by default

Reply:
Yes, we want this to be a Dask sparse array. How to convert from dask.bag to dask sparse array is one point I have not figured out yet.

Reply:
Sorry for the delay in answering @rth!

I think the best way forward is something like

```python
bag2 = X.map_partitions(transformer)

objs = bag2.to_delayed()
arrs = [da.from_delayed(obj, (np.nan, self.n_features), self.dtype)
        for obj in objs]
result = da.concatenate(arrs, axis=0)
result
# dask.array<concatenate, shape=(nan, 1048576), dtype=float64, chunksize=(nan, 1048576)>
```

there are two potential issues with that approach

1. We don't know the length of each partition / block, so we end up with an array that has unknown chunks. This can cause issues with later estimators, but that's a larger issue than we should worry about for this PR.
2. `result.compute()` isn't actually valid :)

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-43-a4f886589eb2> in <module>()
----> 1 result.compute()
ValueError: setting an array element with a sequence.
```

That's because while you can store a `scipy.sparse` matrix in a dask.array, not everything works correctly. I have a WIP OneHotEncoder PR that runs into the same issue. I think for know, that's OK, it'll need to be solved in dask (or maybe by dask-ml by using `pydata/sparse`). But again, that's something we need to figure out library-wide.

Reply:
I suspect that short term the solution for sparse arrays will be to place scipy.sparse matrices within dask arrays, but then avoiding using dask.array algorithms on those arrays.  Dask arrays should be fine as *containers* for sparse matrices, even if the algorithms don't work.

Note that this quasi-support is relatively new.  See https://github.com/dask/dask/pull/3738

Reply:
> I suspect that short term the solution for sparse arrays will be to place scipy.sparse matrices within dask arrays

Agreed, short-term that's probably best (my phrasing above did not indicate that, my apologies).

Reply:
I think in the example, it would be best to note this and show some workarounds

```python
import sparse

X.map_blocks(sparse.COO.from_scipy_sparse, dtype=X.dtype).compute()
```

@rth do you have time to make these changes? Sorry again for letting this linger.